bundles/org.eclipse.jface.text/projection/org/eclipse/jface/text/source/projection/ProjectionViewer.java

```diff
@@ -272,6 +272,34 @@ private void computeExpectedExecutionCosts() {
 		}
 	}
 
+	/**
+	 * An {@link IDocumentListener} that makes sure that {@link #fVisibleRegionDuringProjection} is
+	 * updated when the document changes and ensures that the collapsed region after the visible
+	 * region is recreated appropriately.
+	 */
+	private final class UpdateDocumentListener implements IDocumentListener {
+		@Override
+		public void documentChanged(DocumentEvent event) {
+			if (fVisibleRegionDuringProjection != null) {
+				int oldLength= event.getLength();
+				int newLength= event.getText().length();
+				int oldVisibleRegionEnd= fVisibleRegionDuringProjection.getOffset() + fVisibleRegionDuringProjection.getLength();
```

Comment:
I haven't tried in editor yet, just curious what happens if the entire element is deleted, either in text or outside (like "delete method" in the outline view).

Reply:
If it is deleted in the editor, it results in the element being removed from the editor (and the editor being empty) which is the same as before without any issues.
If it is deleted outside, it works the same _most_ of the time but there is one edge case (if it is at the end of the file and the method is "big enough") where you can select the surrounding element and the closing `}` is missing. I'll take a look at that.

Reply:
That edge case was an off-by-1 error that should now be fixed by changing `event.getOffset() + oldLength < oldVisibleRegionEnd` condition to `event.getOffset() + oldLength <= oldVisibleRegionEnd`)